engineapi/backups.go

```diff
@@ -42,17 +41,36 @@ func getBackupCredentialEnv(backupTarget string, credential map[string]string) (
 		return envs, err
 	}
 	if backupType == types.BackupStoreTypeS3 {
-		if credential != nil && credential[types.AWSAccessKey] != "" && credential[types.AWSSecretKey] != "" {
-			envs = append(envs, fmt.Sprintf("%s=%s", types.AWSAccessKey, credential[types.AWSAccessKey]))
-			envs = append(envs, fmt.Sprintf("%s=%s", types.AWSSecretKey, credential[types.AWSSecretKey]))
+		if credential != nil {
+			if credential[types.AWSIAMRole] == "" {
+				if credential[types.AWSAccessKey] == "" && credential[types.AWSSecretKey] == "" {
+					return nil, fmt.Errorf("Could not backup to %s without setting credential", backupType)
+				}
+				if credential[types.AWSAccessKey] == "" && credential[types.AWSSecretKey] != "" {
+					return nil, fmt.Errorf("Could not backup to %s without setting credential access key", backupType)
+				}
+				if credential[types.AWSAccessKey] != "" && credential[types.AWSSecretKey] == "" {
+					return nil, fmt.Errorf("Could not backup to %s without setting credential secret access key", backupType)
+				}
+				// Access S3 by AWS credentials
+				if credential[types.AWSAccessKey] != "" && credential[types.AWSSecretKey] != "" {
+					envs = append(envs, fmt.Sprintf("%s=%s", types.AWSAccessKey, credential[types.AWSAccessKey]))
+					envs = append(envs, fmt.Sprintf("%s=%s", types.AWSSecretKey, credential[types.AWSSecretKey]))
+				}
+			} else {
+				// Access S3 by AWS IAM role, and configure credentials if present
+				if credential[types.AWSAccessKey] != "" && credential[types.AWSSecretKey] != "" {
+					envs = append(envs, fmt.Sprintf("%s=%s", types.AWSAccessKey, credential[types.AWSAccessKey]))
+					envs = append(envs, fmt.Sprintf("%s=%s", types.AWSSecretKey, credential[types.AWSSecretKey]))
+				}
+			}
+
 			envs = append(envs, fmt.Sprintf("%s=%s", types.AWSEndPoint, credential[types.AWSEndPoint]))
 			envs = append(envs, fmt.Sprintf("%s=%s", types.AWSCert, credential[types.AWSCert]))
 			envs = append(envs, fmt.Sprintf("%s=%s", types.HTTPSProxy, credential[types.HTTPSProxy]))
 			envs = append(envs, fmt.Sprintf("%s=%s", types.HTTPProxy, credential[types.HTTPProxy]))
 			envs = append(envs, fmt.Sprintf("%s=%s", types.NOProxy, credential[types.NOProxy]))
 			envs = append(envs, fmt.Sprintf("%s=%s", types.VirtualHostedStyle, credential[types.VirtualHostedStyle]))
-		} else if os.Getenv(types.AWSAccessKey) == "" || os.Getenv(types.AWSSecretKey) == "" {
```

Comment:
There seems to be a change in behavior, can you verify that we don't need to support this anymore ?

Since previously if they where already present in the env, we would proceed, now it looks like the passing of a credential is required.
Which seems reasonable, since we are also setting proxy envs, but needs to be verified that this is desired.

Reply:
I find this https://longhorn.io/docs/1.1.0/references/examples/#restore-to-file
It'll read the environment variables `AWS_ACCESS_KEY_ID` & `AWS_SECRET_ACCESS_KEY` and `AWS_ENDPOINTS`.

The helm chart or longhorn-manager itself won't expose `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` environment variables.